ufl/algorithms/apply_derivatives.py

```diff
@@ -1057,25 +1276,57 @@ def __init__(self, coefficients, arguments, coefficient_derivatives):
             coefficient_derivatives=coefficient_derivatives,
         )
 
-    # Explicitly defining dg/dw == 0
-    geometric_quantity = GenericDerivativeRuleset.independent_terminal
+    # Work around singledispatchmethod inheritance issue;
+    # see https://bugs.python.org/issue36457.
+    @singledispatchmethod
+    def process(self, o: Expr) -> Expr:
+        """Process ``o``.
 
-    def cell_avg(self, o, fp):
+        Args:
+            o: `Expr` to be processed.
+
+        Returns:
+            Processed object.
+
+        """
+        return super().process(o)
+
+    # --- Specialized rules for geometric quantities
+
+    @process.register(GeometricQuantity)
+    def _(self, o: Expr) -> Expr:
+        # Explicitly defining dg/dw == 0
+        return self.independent_terminal(o)
+
+    @process.register(CellAvg)
+    @DAGTraverser.postorder
+    def _(self, o: Expr, fp: Expr) -> Expr:
         """Differentiate a cell_avg."""
         # Cell average of a single function and differentiation
         # commutes, D_f[v](cell_avg(f)) = cell_avg(v)
         return cell_avg(fp)
 
-    def facet_avg(self, o, fp):
+    @process.register(FacetAvg)
+    @DAGTraverser.postorder
+    def _(self, o: Expr, fp: Expr) -> Expr:
         """Differentiate a facet_avg."""
         # Facet average of a single function and differentiation
         # commutes, D_f[v](facet_avg(f)) = facet_avg(v)
         return facet_avg(fp)
 
-    # Explicitly defining da/dw == 0
-    argument = GenericDerivativeRuleset.independent_terminal
+    @process.register(Argument)
+    def _(self, o: Expr) -> Expr:
+        # Explicitly defining da/dw == 0
+        return self._process_argument(o)
 
-    def coefficient(self, o):
+    def _process_argument(self, o: Expr) -> Expr:
+        return self.independent_terminal(o)
+
+    @process.register(Coefficient)
+    def _(self, o: Expr) -> Expr:
+        return self._process_coefficient(o)
+
+    def _process_coefficient(self, o: Expr) -> Expr:
```

Comment:
Is the doc-string here strictly correct? (I know it existed prior to the PR).
As `_process_coefficient` supports taking a `BaseFormOperator` or a `Cofunction` it is not really a `Coefficient`.

I guess this goes back to the `GateauxDerivativeRuleset` where any `ExprList` can be taken as input and therefore might be out of scop of this PR.

Reply:
I partially fixed this (as I added these new methods), but I think we should do more comprehensive fix in a future PR.